README.md
# Threads.

<img height="300" src="client/assets/banner.png"></img>

## Description:
The project aims to clone the Threads from Instagram social network in order to understand how this type of application could work, using Flutter framework & Firebase Database.
This repository is use for educationnal content.

## 🚧 The project if not finished.

The project is under construction, activly developed, `feel free to star the project` if you want to support me and be notify to the major update of the construction of this social network. This just the begining be patient!

## 💡 Idea

if you want to see in realtime the progression of the evolution of project, follow the link: https://github.com/users/Antoinegtir/projects/6

## 🔗 Link of my article about the implemtation

 <a href="https://medium.com/@zkhwctb/how-i-created-an-open-source-threads-clone-with-flutter-bddc7b6ebc55">
    <img src="https://img.shields.io/badge/medium-fff?style=for-the-badge&logo=medium&logoColor=black" alt="Medium">
</a>

## Screenshots



Follow Page         |  Privacy Page       |   Threads Page               |  Feed Page
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/onboard.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/privacy.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/threads.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/feed.png?raw=true)|

Profile Page         |  Search Page       |   Notification Page               |  Edit Page
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/profile.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/search.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/notification.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/edit.png?raw=true)|

Settings Page         |  Post Page       |   Crop Page               |  Feed Page
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/settings.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/post.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/crop.png?raw=true)|![](https://github.com/Antoinegtir/instagram-threads/blob/main/screenshot/feed.png?raw=true)|

## Language & Tools used:

<img src="https://skillicons.dev/icons?i=flutter,dart,firebase"/>

## Aivailable on:
iOS, MacOS, Android, Web.

## Usage

- install flutter engine -> https://docs.flutter.dev/get-started/install
- tap command `flutter run` in the root folder of the project

## Author

@Antoinegtir


CHANGELOG.md
commit 863448ef27cff96083e57f856681151e3daf9984
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 07:14:31 2023 +0200

    ⚙️ update state manager

commit 24178e9bca777e031bc3aa4bc22cc11bca96a47b
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 07:14:04 2023 +0200

    📩 update graphical interface

commit cf72ee514931e35175631ab883d8a4a30c08c84b
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:42:24 2023 +0200

    📩 add post graphical interface

commit b2c284344f597b3dcb1207ae338c0bd657ce2a37
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:41:45 2023 +0200

    📭 add compositing state manager

commit bbb44b1ded1116a1eedcf47105eb5dee320fdcfd
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:41:20 2023 +0200

    🙎‍♂️ update profile page

commit 7c8acbba13826500b49b710621d3a3ef0f57c2ba
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:40:41 2023 +0200

    🧪 remove unit test

commit cc8006cdce1887107fff920514cdf7dc21dec77d
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:40:16 2023 +0200

    🌍 update platform transformers

commit 08b1e4866b859b84c5f31e20b5e13b8922ba3bcd
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:39:27 2023 +0200

    🙈 update ignore file

commit 52483d583756c8800dd342288ac542adaff28e2d
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:39:13 2023 +0200

    🙈 update ignore file

commit 11c0d3d989ce2157ae23974801d3d030f4b14bc4
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:38:45 2023 +0200

    🏠 update homepage

commit caeb6e4861f8942d13cb484357875fde8863ca81
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:37:20 2023 +0200

    ⚙️ update model stucture

commit 7f6a3dc080c394f7b8ffa8842e9c5aba9de3c024
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:36:38 2023 +0200

    ➖ remove comentary

commit 1c51c66fd77ef85f2b0bd92367e3e069020d97f2
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Fri Jul 21 06:36:00 2023 +0200

    🏃‍♀️ add animation route transition

commit 1cd358f60948539c11dd4ed865392e0de277ac76
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Wed Jul 19 21:04:08 2023 +0200

    :memo: add readme

commit ac7a981488066c425cac74d3cf23babbfbd22d99
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Wed Jul 19 10:02:37 2023 +0200

    :sparkles: add tabbar

commit 005df828b2809e7146e50f917b8fa7ab165c8ddc
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Wed Jul 19 10:02:08 2023 +0200

    :bento: change style

commit 2af2acb8b97dcc9805e341daa1605e6483e2a264
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Wed Jul 19 08:55:38 2023 +0200

    :bento: add assets

commit a1896dd174e069ab09546069fdc040f6e0691277
Author: Antoine • Developer <114834504+Antoinegtir@users.noreply.github.com>
Date:   Mon Jul 17 23:57:40 2023 +0200

    Update README.md

commit 1bddd4d6aab610a639c67157fd38d7f6212f98db
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Mon Jul 17 23:56:39 2023 +0200

    :bento: update onboard

commit 4b0c64abd209366530163d18e9776ac1036ccd45
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:25:11 2023 +0200

    :bento: update onboard

commit c66bb44e8845586bd51a16e60f9cc28751c87651
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:19:22 2023 +0200

    :bento: update onboard

commit 067926af20c3444a23928cd29c9587c6a02092a9
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:19:01 2023 +0200

    :bento: update onboard

commit 66c075d6a47400c4db9ca6628327a7f84be4b22c
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:18:25 2023 +0200

    :bento: update onboard

commit d173b4a2d0426a6252ca227943e629c4e14da4c4
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:17:40 2023 +0200

    :bento: update onboard

commit 73042df4e354579807012e6a35fbf64d8efbfc2f
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:15:35 2023 +0200

    :bento: update onboard

commit 30b0991ab7ed6a046e75d9cabf182fb4e939b258
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:15:03 2023 +0200

    :bento: update onboard

commit b71d96a31e5dc0f0615f1e1401e125e1a4018f3f
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:13:33 2023 +0200

    :bento: update onboard

commit 78390ca4319d5707204136c0df6bd33e3916b037
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 22:13:13 2023 +0200

    :bento: update onboard

commit 76c2bb571add7aca0a2dabdd8e2f02765fa900d4
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 21:55:21 2023 +0200

    :memo: update readme

commit 2336b1354a6f342d819e5763fc3b7d9758f51201
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 21:53:54 2023 +0200

    :sparkles: update readm

commit 1444192e61c5f96ea97ed0e81ffa64ddd3e033f2
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 21:53:09 2023 +0200

    :sparkles: update signin

commit 675d8c586d47910da51d35d617be080d4fde347b
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 21:22:23 2023 +0200

    :sparkles: update readme

commit 41bc04a8de20a5041af1572bd37d6fef3a6d51fe
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 15:47:09 2023 +0200

    ✨ booleean that detect if it is my profile

commit 954c2fadc72d40af1eae84a4d941a42649aeff58
Author: Antoine • Developer <antoine.gonthier@epitech.eu>
Date:   Sun Jul 16 15:20:56 2023 +0200

    ⚙️add condition for firebase


client/lib/auth/signup.dart
import 'dart:io';

import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:image_cropper/image_cropper.dart';
import 'package:image_picker/image_picker.dart';
import 'package:threads/auth/email.dart';

class Signup extends StatefulWidget {
  const Signup({Key? key}) : super(key: key);
  @override
  State<StatefulWidget> createState() => _SignupState();
}

class _SignupState extends State<Signup> {
  late TextEditingController _nameController;
  late TextEditingController _bioController;
  late TextEditingController _linkController;
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  @override
  void initState() {
    _nameController = TextEditingController();
    _bioController = TextEditingController();
    _linkController = TextEditingController();
    super.initState();
  }

  @override
  void dispose() {
    _nameController.dispose();
    _bioController.dispose();
    _linkController.dispose();
    super.dispose();
  }

  File? _image;

  Future<void> getImage(BuildContext context, ImageSource source,
      Function(File) onImageSelected) async {
    ImagePicker()
        .pickImage(source: source, imageQuality: 100)
        .then((XFile? file) async {
      await ImageCropper.platform.cropImage(
        sourcePath: file!.path,
        cropStyle: CropStyle.circle,
        aspectRatioPresets: [
          CropAspectRatioPreset.square,
          CropAspectRatioPreset.ratio3x2,
          CropAspectRatioPreset.original,
          CropAspectRatioPreset.ratio4x3,
          CropAspectRatioPreset.ratio16x9
        ],
      ).then((value) => setState(() {
            onImageSelected(File(value!.path));
          }));
    });
  }

  Widget _body(BuildContext context) {
    return Column(mainAxisAlignment: MainAxisAlignment.start, children: [
      Container(
        height: 120,
      ),
      Text(
        "Profile",
        style: TextStyle(
            color: Colors.white, fontSize: 28, fontWeight: FontWeight.w600),
        textAlign: TextAlign.center,
      ),
      Container(
        height: 20,
      ),
      Text(
        "Customize your Threads profile.",
        style: TextStyle(
            color: Color.fromARGB(255, 96, 96, 96),
            fontSize: 16,
            fontWeight: FontWeight.w600),
        textAlign: TextAlign.center,
      ),
      Container(
        height: 100,
      ),
      Container(
          height: 250,
          width: 330,
          decoration: BoxDecoration(
            color: Color.fromARGB(255, 25, 25, 25),
            borderRadius: BorderRadius.circular(20),
            border: Border.all(
              color: Colors.grey,
              width: 0.5,
            ),
          ),
          child: Padding(
              padding: EdgeInsets.symmetric(horizontal: 20),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Container(
                        width: 200,
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Container(
                              height: 10,
                            ),
                            Text(
                              "Name",
                              style: TextStyle(
                                  color: Colors.white,
                                  fontWeight: FontWeight.w600,
                                  fontSize: 18),
                            ),
                            CupertinoTextField(
                              controller: _nameController,
                              prefix: Icon(
                                Icons.lock_outline_rounded,
                                size: 15,
                                color: Colors.white,
                              ),
                              style:
                                  TextStyle(color: Colors.white, fontSize: 18),
                              placeholder: 'Antoine.gtir',
                              placeholderStyle:
                                  TextStyle(color: Colors.grey, fontSize: 18),
                              padding: EdgeInsets.all(8),
                              decoration: BoxDecoration(
                                borderRadius: BorderRadius.circular(8),
                              ),
                            ),
                            Container(
                              height: 5,
                            ),
                            Container(
                              width: 300,
                              height: 0.5,
                              color: Colors.grey,
                            ),
                            Container(
                              height: 20,
                            ),
                          ],
                        ),
                      ),
                      Container(
                        width: 30,
                      ),
                      GestureDetector(
                        onTap: () {
                          showCupertinoModalPopup(
                              context: context,
                              builder: (BuildContext context) => CupertinoTheme(
                                    data: CupertinoThemeData(
                                      brightness: Brightness
                                          .dark, // Définir le mode sombre
                                    ),
                                    child: CupertinoActionSheet(
                                      title: Text('Changer de photo de profil'),
                                      message: Text(
                                          'Ta photo de profil est visible par tous et permetttra à tes amis de t\'ajoyter plus facilement'),
                                      actions: <Widget>[
                                        CupertinoActionSheetAction(
                                          child: Text('Photothèque'),
                                          onPressed: () {
                                            getImage(
                                                context, ImageSource.gallery,
                                                (file) {
                                              setState(() {
                                                _image = file;
                                              });
                                            });
                                            setState(() {});
                                            Navigator.pop(context);
                                          },
                                        ),
                                        CupertinoActionSheetAction(
                                          child: Text('Appareil photo'),
                                          onPressed: () {
                                            getImage(
                                                context, ImageSource.camera,
                                                (file) {
                                              setState(() {
                                                _image = file;
                                              });
                                            });
                                            Navigator.pop(context);
                                          },
                                        ),
                                        CupertinoActionSheetAction(
                                          child: Text(
                                            'Supprimer la photo de profil',
                                            style: TextStyle(color: Colors.red),
                                          ),
                                          onPressed: () {
                                            Navigator.pop(context);
                                          },
                                        ),
                                      ],
                                      cancelButton: CupertinoActionSheetAction(
                                        child: Text('Annuler'),
                                        onPressed: () {
                                          Navigator.pop(context);
                                        },
                                      ),
                                    ),
                                  ));
                        },
                        child: CircleAvatar(
                            backgroundColor: Colors.transparent,
                            radius: 25,
                            backgroundImage: (_image != null
                                ? FileImage(_image!)
                                : CachedNetworkImageProvider(
                                    scale: 10,
                                    "https://firebasestorage.googleapis.com/v0/b/threads-instagram.appspot.com/o/Simulator%20Screen%20Shot%20-%20iPhone%2014%20Pro%20-%202023-07-09%20at%2021.28.44.jpeg?alt=media&token=9ba68508-7555-4b64-824c-e17ee14753a3",
                                  ) as ImageProvider)),
                      )
                    ],
                  ),
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        "Bio",
                        style: TextStyle(
                            color: Colors.white,
                            fontWeight: FontWeight.w600,
                            fontSize: 18),
                      ),
                      CupertinoTextField(
                        controller: _bioController,
                        prefix: Icon(
                          Icons.add,
                          size: 15,
                          color: Colors.white,
                        ),
                        style: TextStyle(color: Colors.white, fontSize: 18),
                        placeholder: 'Write bio',
                        placeholderStyle:
                            TextStyle(color: Colors.grey, fontSize: 16),
                        padding: EdgeInsets.all(8),
                        decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                      Container(
                        height: 10,
                      ),
                      Container(
                        width: 300,
                        height: 0.5,
                        color: Colors.grey,
                      ),
                      Container(
                        height: 20,
                      ),
                    ],
                  ),
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        "Link",
                        style: TextStyle(
                            color: Colors.white,
                            fontWeight: FontWeight.w600,
                            fontSize: 18),
                      ),
                      CupertinoTextField(
                        controller: _linkController,
                        prefix: Icon(
                          Icons.add,
                          size: 15,
                          color: Colors.white,
                        ),
                        style: TextStyle(color: Colors.white, fontSize: 18),
                        placeholder: 'Add Link',
                        placeholderStyle:
                            TextStyle(color: Colors.grey, fontSize: 16),
                        padding: EdgeInsets.all(8),
                        decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                    ],
                  ),
                ],
              ))),
      Container(
        height: 20,
      ),
      GestureDetector(
          onTap: () {
            Navigator.push(
                context,
                MaterialPageRoute(
                    builder: (context) => EmailPage(
                          name: _nameController.text,
                          bio: _bioController.text,
                          link: _linkController.text,
                          image: _image,
                        )));
          },
          child: Container(
            height: 50,
            width: MediaQuery.of(context).size.width - 80,
            decoration: BoxDecoration(
                color: Colors.white, borderRadius: BorderRadius.circular(20)),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Icon(
                  CupertinoIcons.square_arrow_down,
                  color: Colors.black,
                ),
                Text(
                  " Import from Instagram",
                  style: TextStyle(color: Colors.black, fontSize: 16),
                )
              ],
            ),
          ))
    ]);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      extendBodyBehindAppBar: true,
      key: _scaffoldKey,
      backgroundColor: Colors.black,
      appBar: AppBar(
        leading: Container(),
        flexibleSpace: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Column(
              children: [
                Container(
                  height: 50,
                ),
                Row(
                  children: [
                    Stack(
                      children: [
                        BackButton(),
                        GestureDetector(
                          onTap: () {
                            Navigator.pop(context);
                          },
                          child: Padding(
                              padding: EdgeInsets.only(left: 35, top: 12),
                              child: Text("Back",
                                  style: TextStyle(
                                    color: Colors.white,
                                    fontSize: 20,
                                  ))),
                        )
                      ],
                    ),
                  ],
                )
              ],
            ),
            Column(
              children: [
                Container(
                  height: 50,
                ),
                Row(
                  children: [
                    Padding(
                        padding: EdgeInsets.only(right: 10, top: 12),
                        child: Stack(
                          children: [
                            GestureDetector(
                              onTap: () {
                                Navigator.push(
                                    context,
                                    MaterialPageRoute(
                                        builder: (context) => EmailPage()));
                              },
                              child: Text("Skip",
                                  style: TextStyle(
                                    color: Colors.white,
                                    fontSize: 20,
                                  )),
                            ),
                            Padding(
                                padding: EdgeInsets.only(left: 42),
                                child: Icon(Icons.arrow_forward_ios_rounded))
                          ],
                        )),
                  ],
                )
              ],
            ),
          ],
        ),
        backgroundColor: Colors.black,
      ),
      body: SingleChildScrollView(child: _body(context)),
    );
  }
}
